### DISCUSSION TITLE
What does line['dir'] identify / How to add tilted/rotated text on pdf page

### DISCUSSION THREAD

**Discussion by rafayaar** [pymupdf/PyMuPDF#3946] [General]
Hi @JorjMcKie 

### Point 1
I am extracting a pdf where I have rotated text. I have extracted the text using get_text method. I assume line['dir'] that I am getting is telling about direction/rotation of text. Please confirm how to use it

### Point 2
After extraction, I would want to add that text to some new pdf. And would want to add as according to original pdf. I think Matrix is to be used to implement this. Not sure!

Please help!

**Comment by JorjMcKie** [comment]
The `line["dir"]` is a tuple of floats with the meaning `(cosine, sine)` of the angle which the line has with the x-axis.

You can view it as representing the rotation of the respective text around the insertion point (`span["origin"]`) of the first span of the line.

There are functions available in `pymupdf` which allow recovering the `Quad` of the line. In `get_text("dict")`, only bboxes are returned: for non-horizontal text the bbox does not tightly wrap the line text. This is shown in [this](https://pymupdf.readthedocs.io/en/latest/_images/img-quads.jpg) picture in the documentation.

If you want to insert some text with the same angle on some other page, the simplest way is to use `page.insert_text(point, text, ..., morph=(point, matrix))`. Here, `point` is the origin of the first span, text is the line text. Use the right (or desired) values for fontname, font size, color etc.

But most importantly define `matrix = pymupdf.Matrix(cosine, -sine, sine, cosine, 0, 0)` - using the `line["dir"]` values.

**Reply by rafayaar** [reply]
Thanks. It solved my problem

I think such examples should be in documentation. Thanks @JorjMcKie . This is the second problem you solved for me. Will be asking more if needed :)

**Reply by JorjMcKie** [reply]
You are welcome anytime 😎!
